From 51638c6a674409de0f1cf1735eeadada6d158f7a Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Thu, 28 Jul 2022 05:34:36 +0300
Subject: [PATCH] [SW-94864] rom: extend address check api

Address check api is extended to support FW SRAM range. This
is to be used as a pre-condition to validate possible load
address range for dma.

Change-Id: I5e435055ec31a47e67b860e8f584c48c5a81e6ca
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 2246c3d..41d45e5 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -655,6 +655,11 @@
 
 	arcboot_get_boot_desc(img, desc, upd_status);
 
+	if (!arcboot_is_valid_address_range(desc->loadaddr, desc->img_size)) {
+		arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_LOADADDR_INVALID);
+		return ARCBOOT_BRINGUP_ERROR;
+	}
+
 	security_report_img_fw(arcboot_rom_global.curr_img, desc->sec_ver);
 
 	/*
@@ -1006,6 +1011,14 @@
 	if (addr >= HL_SRAM_BASE_FULL && addr + len <= HL_SRAM_BASE_FULL + HL_SRAM_SIZE)
 		return true;
 
+	if (addr >= HL_FW_MEM_BASE + HL_FWSRAM_OFFSET &&
+	    addr + len <= HL_FW_MEM_BASE + HL_FWSRAM_OFFSET + HL_FW_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_FW_SRAM_FULL_BASE &&
+	    addr + len <= HL_FW_SRAM_FULL_BASE + HL_FW_SRAM_SIZE)
+		return true;
+
 	return false;
 }
 
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index b14f90e..e3b1006 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -42,110 +42,115 @@
 	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 35,
 	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 36,
 	PROGRESS_BRINGUP_PCI_PRI = 37,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 38,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 39,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 40,
-	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 41,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 42,
-	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 43,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 44,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 45,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 46,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 47,
-	PROGRESS_BRINGUP_PCI_SEC = 48,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 49,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 50,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 51,
-	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 52,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 53,
-	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 54,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 55,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 56,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 57,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 58,
-	PROGRESS_ARCBOOT_IN_RECOVERY = 59,
-	PROGRESS_ARCBOOT_ROM_TPM = 60,
-	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 61,
-	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 62,
-	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 63,
-	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 64,
-	PROGRESS_BRINGUP_TPM_SPI = 65,
-	PROGRESS_BRINGUP_TPM_SPI_PRI = 66,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 67,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 68,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 69,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 70,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 71,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 72,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 73,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 74,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 75,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 76,
-	PROGRESS_BRINGUP_TPM_SPI_SEC = 77,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 78,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 79,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 80,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 81,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 82,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 83,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 84,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 85,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 86,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 87,
-	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 88,
-	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 89,
-	PROGRESS_PPBOOT_START = 90,
-	PROGRESS_PPBOOT_FINISH = 91,
-	PROGRESS_BRINGUP_RAM_PRI = 92,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 93,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 94,
-	PROGRESS_BRINGUP_RAM_SEC = 95,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 96,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 97,
-	PROGRESS_ARCBOOT_RAM_START = 98,
-	PROGRESS_ARCBOOT_RECOVERY = 99,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 100,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 101,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 102,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 103,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 104,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 105,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 106,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 107,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 108,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 109,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 110,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 111,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 112,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 113,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 114,
-	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 115,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 116,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 117,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 118,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 119,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 120,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 121,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 122,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 123,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 124,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 125,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 126,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 127,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 128,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 129,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 130,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 131,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 132,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 133,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 134,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 135,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 136,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 137,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 138,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 139,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 140,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 141,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 142,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 143,
+	PROGRESS_BRINGUP_PCI_PRI_LOADADDR_INVALID = 38,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 39,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 40,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 41,
+	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 42,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 43,
+	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 44,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 45,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 46,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 47,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 48,
+	PROGRESS_BRINGUP_PCI_SEC = 49,
+	PROGRESS_BRINGUP_PCI_SEC_LOADADDR_INVALID = 50,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 51,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 52,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 53,
+	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 54,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 55,
+	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 56,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 57,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 58,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 59,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 60,
+	PROGRESS_ARCBOOT_IN_RECOVERY = 61,
+	PROGRESS_ARCBOOT_ROM_TPM = 62,
+	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 63,
+	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 64,
+	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 65,
+	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 66,
+	PROGRESS_BRINGUP_TPM_SPI = 67,
+	PROGRESS_BRINGUP_TPM_SPI_PRI = 68,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_LOADADDR_INVALID = 69,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 70,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 71,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 72,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 73,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 74,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 75,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 76,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 77,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 78,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 79,
+	PROGRESS_BRINGUP_TPM_SPI_SEC = 80,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_LOAD_ADDR_INVALID = 81,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 82,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 83,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 84,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 85,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 86,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 87,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 88,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 89,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 90,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 91,
+	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 92,
+	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 93,
+	PROGRESS_PPBOOT_START = 94,
+	PROGRESS_PPBOOT_FINISH = 95,
+	PROGRESS_BRINGUP_RAM_PRI = 96,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 97,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 98,
+	PROGRESS_BRINGUP_RAM_SEC = 99,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 100,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 101,
+	PROGRESS_ARCBOOT_RAM_START = 102,
+	PROGRESS_ARCBOOT_RECOVERY = 103,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 104,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 105,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 106,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 107,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 108,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 109,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 110,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 111,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 112,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 113,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 114,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 115,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 116,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 117,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 118,
+	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 119,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 120,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 121,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 122,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 123,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 124,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 125,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 126,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 127,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 128,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 129,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 130,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 131,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 132,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 133,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 134,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 135,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 136,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 137,
+	PROGRESS_ARCBOOT_EROM_IMG0_LOADADDR_INVALID = 138,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 139,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 140,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 141,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 142,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 143,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 144,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 145,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 146,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 147,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 148,
 };
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 68afffa..8e86f49 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -51,6 +51,11 @@
 	if (idx == HL_FW_IMG_NOT_FOUND)
 		return;
 
+	if (!arcboot_is_valid_address_range(desc.loadaddr, desc.img_size)) {
+		arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_LOADADDR_INVALID);
+		return;
+	}
+
 	dst_type = arcboot_rom_get_dma_dest_type(desc.loadaddr);
 
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_DMA_START);
diff --git a/include/arcboot.h b/include/arcboot.h
index 880baee..4a29831 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -95,6 +95,7 @@
 
 enum arcboot_rom_bringup_proggress {
 	PROGRESS_BRINGUP_START,
+	PROGRESS_BRINGUP_LOADADDR_INVALID,
 	PROGRESS_BRINGUP_TOO_BIG,
 	PROGRESS_BRINGUP_CRC_FAIL,
 	PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR,
@@ -132,6 +133,7 @@
 
 enum arcboot_rom_erom_img_progress {
 	PROGRESS_EROM_BRINGUP_START,
+	PROGRESS_EROM_LOADADDR_INVALID,
 	PROGRESS_EROM_DMA_START,
 	PROGRESS_EROM_DMA_CRC_FAILED,
 	PROGRESS_EROM_DMA_UNALIGNED_DEST_ADDR,
